src/gpt.c

```diff
@@ -0,0 +1,346 @@
+		g_propagate_error(error, ierror);
+		goto out_deassign;
+	}
+
+	if (fdisk_get_partition(cxt, BOOT_PARTITION_ENTRY, &pa) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"No boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out_deassign;
+	}
+
+	sector_size = fdisk_get_sector_size(cxt);
+	if ((region_start / sector_size) ==
+	    fdisk_partition_get_start(pa)) {
+		partition->start = region_start + region_size / 2;
+	} else if (((region_start + region_size / 2) / sector_size) ==
+	           fdisk_partition_get_start(pa)) {
+		partition->start = region_start;
+	} else {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Boot partition's start address does not match "
+				"region configuration");
+		goto out_unref_part;
+	}
+	partition->size = region_size / 2;
+
+	res = TRUE;
+
+out_unref_part:
+	fdisk_unref_partition(pa);
+out_deassign:
+	fdisk_deassign_device(cxt, 0);
+out_cxt:
+	fdisk_unref_context(cxt);
+
+	return res;
+}
+
+gboolean r_gpt_switch_set_boot_partition(const gchar *device,
+		const struct boot_switch_partition *partition,
+		GError **error)
+{
+	gboolean res = FALSE;
+	GError *ierror = NULL;
+	gulong sector_size;
+	struct fdisk_context *cxt = NULL;
+	struct fdisk_label *lb = NULL;
+	struct fdisk_partition *pa = NULL;
+
+	g_return_val_if_fail(device, FALSE);
+	g_return_val_if_fail(partition, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	cxt = get_context();
+
+	if (fdisk_assign_device(cxt, device, 0)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to open %s with libfdisk", device);
+		goto out_cxt;
+	}
+
+	if (!check_gpt(cxt, &ierror)) {
+		g_propagate_error(error, ierror);
+		goto out_deassign;
+	}
+
+	lb = fdisk_get_label(cxt, NULL);
+	if (!lb)
+		g_error("%s: Failed to get libfdisk label\n", G_STRLOC);
+
+	if (fdisk_label_is_changed(lb) == 1) {
+		g_message("GPT is inconsistent, repairing...\n");
+		if (fdisk_write_disklabel(cxt) != 0) {
+			g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+					"Could not repair GPT");
+			goto out_deassign;
+		}
+		g_message("GPT repaired\n");
+	}
+	if (fdisk_verify_disklabel(cxt) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Old GPT failed to verify");
+		goto out_deassign;
+	}
+
+	if (fdisk_get_partition(cxt, BOOT_PARTITION_ENTRY, &pa) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"No boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out_deassign;
+	}
+
+	sector_size = fdisk_get_sector_size(cxt);
+	fdisk_reset_partition(pa); /* only change the location */
+	fdisk_partition_set_start(pa, partition->start / sector_size);
+	fdisk_partition_set_size(pa, partition->size / sector_size);
+
+	if (fdisk_set_partition(cxt, BOOT_PARTITION_ENTRY, pa) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not update boot partition");
+		goto out_deassign;
```

Comment:
Here the partition `pa` is already a valid reference so it has to be unreferenced:
```suggestion
		goto out_unref_part;
```